scripts/exglobal_diag.sh

```diff
@@ -176,100 +165,100 @@ EOFdiag
                numfile[n]=$(expr ${numfile[n]} + 1)
-            elif [ $count -eq 1 ]; then
-                cat ${prefix}${type}_${loop}* > diag_${type}_${string}.${CDATE}${DIAG_SUFFIX}
-                if [ $DIAG_COMPRESS = "YES" ]; then
-                   $COMPRESS diag_${type}_${string}.${CDATE}${DIAG_SUFFIX}
+            elif [[ "${count}" -eq 1 ]]; then
+                cat "${prefix}${type}_${loop}"* > "diag_${type}_${string}.${CDATE}${DIAG_SUFFIX}"
+                if [[ "${DIAG_COMPRESS}" = "YES" ]]; then
+                   ${COMPRESS} "diag_${type}_${string}.${CDATE}${DIAG_SUFFIX}"
                 fi
                 echo "diag_${type}_${string}.${CDATE}*" >> ${diaglist[n]}
-                numfile[n]=$(expr ${numfile[n]} + 1)
+                numfile[n]=$(expr "${numfile[n]}" + 1)
             fi
          done
       done
-      echo $(date) END loop $string >&2
+      echo $(date) END loop "${string}" >&2
    done
 
    # We should already be in $DATA, but extra cd to be sure.
-   cd $DATA
+   cd "${DATA}" || exit
 
    # If requested, compress diagnostic files
-   if [ $DIAG_COMPRESS = "YES" -a $USE_CFP = "NO" ]; then
-      echo $(date) START $COMPRESS diagnostic files >&2
-      for file in $(ls diag_*${CDATE}${DIAG_SUFFIX}); do
-         $COMPRESS $file
+   if [[ "${DIAG_COMPRESS}" = "YES" && "${USE_CFP}" = "NO" ]]; then
+      echo $(date) START "${COMPRESS}" diagnostic files >&2
+      for file in $(ls diag_*"${CDATE}${DIAG_SUFFIX}"); do
+         ${COMPRESS} "${file}"
       done
-      echo $(date) END $COMPRESS diagnostic files >&2
+      echo $(date) END "${COMPRESS}" diagnostic files >&2
    fi
 
-   if [ $USE_CFP = "YES" ] ; then
-      chmod 755 $DATA/mp_diag.sh
-      ncmd=$(cat $DATA/mp_diag.sh | wc -l)
-      if [ $ncmd -gt 0 ]; then
+   if [[ "${USE_CFP}" = "YES" ]] ; then
+      chmod 755 "${DATA}/mp_diag.sh"
+      ncmd=$(cat "${DATA}/mp_diag.sh" | wc -l)
+      if [[ "${ncmd}" -gt 0 ]]; then
          ncmd_max=$((ncmd < max_tasks_per_node ? ncmd : max_tasks_per_node))
-         APRUNCFP_DIAG=$(eval echo $APRUNCFP)
-         $APRUNCFP_DIAG $DATA/mp_diag.sh
+         APRUNCFP_DIAG=$(eval echo "${APRUNCFP}")
+         ${APRUNCFP_DIAG} "${DATA}/mp_diag.sh"
          export err=$?; err_chk
       fi
    fi
 
    # Restrict diagnostic files containing rstprod data
    rlist="conv_gps conv_ps conv_pw conv_q conv_sst conv_t conv_uv saphir"
-   for rtype in $rlist; do
-      for rfile in *"${rtype}"*; do 
+   for rtype in ${rlist}; do
+      for rfile in *"${rtype}"*; do
          if [[ -s "${rfile}" ]]; then
             ${CHGRP_CMD} "${rfile}"
          fi
       done
    done
 
    # If requested, create diagnostic file tarballs
-   if [ $DIAG_TARBALL = "YES" ]; then
+   if [[ ${DIAG_TARBALL} = "YES" ]]; then
       echo $(date) START tar diagnostic files >&2
       n=-1
-      while [ $((n+=1)) -le $ntype ] ;do
+      while [ $((n+=1)) -le ${ntype} ] ;do
```

Comment:
I'm not a fan of doing math on a variable inside of a logical check.  I'd prefer this operation to stand on its own in the loop, but that will change logic and is probably beyond the scope of this PR.  I'll open an issue to handle this kind of thing.  Instead, just suggesting double braces:
```suggestion
      while [[ $((n+=1)) -le ${ntype} ]] ;do
```

Reply:
Opened https://github.com/NOAA-EMC/global-workflow/issues/3653.

Reply:
Sounds good, thanks @DavidHuber-NOAA !